### PR TITLE
Properly ref count diagnostic taggers.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
I noticed a case where diagnostic tags weren't appearing.  I worked a bit and found a repro where we had multiple active taggers handed out, but one was disposed.  Because the taggers were sharing state, this meant that all active taggers were effectively 'disposed'.

I've added a proper refcounting scheme so that we don't actually release the tagger resources if there are outstanding taggers needing them.  

--

I recommend reading this review one commit at a time.

Fixes https://github.com/dotnet/roslyn/issues/4724

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
Reviewreviewers: @balajikris, @basoundr, @brettfo, @davkean, @DustinCampbell, @dpoeschl, @Pilchie, @jasonmalinowski, @heejaechang

**Comment by DustinCampbell** [comment]
Nice catch! I think this is the cause of a problem that @davkean and I saw yesterday. #4724 is filed for that issue. Does it looks like the same thing?

**Comment by CyrusNajmabadi** [comment]
@DustinCampbell Yup.  This is very likely the same issue.  Are you ok with me checking htis in?

File: src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs [hunk 1]
```diff
@@ -23,43 +25,20 @@ public static class SquiggleUtilities
...
             {
-                var service = workspace.Services.GetService<ISolutionCrawlerRegistrationService>() as SolutionCrawlerRegistrationService;
-                service.WaitUntilCompletion_ForTestingPurposesOnly(workspace, ImmutableArray.Create(analyzerService.CreateIncrementalAnalyzer(workspace)));
+                var tagger = wrapper.TaggerProvider.CreateTagger<IErrorTag>(workspace.Documents.First().GetTextBuffer());
+                using (var disposable = tagger as IDisposable)
+                {
+                    wrapper.WaitForTags();
 
-                listener.CreateWaitTask().PumpingWait();
+                    var snapshot = workspace.Documents.First().GetTextBuffer().CurrentSnapshot;
```
Comment (AdamSpeight2008) [inline src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs]: Is this safe to do? What if `Documents` is empty or null?

File: src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs [hunk 2]
```diff
@@ -76,30 +55,20 @@ public abstract class AbstractSquiggleProducerTests
...
+                using (var disposable = tagger as IDisposable)
+                {
+                    source.RaiseDiagnosticsUpdated(updateArgs);
 
-                listener.CreateWaitTask().PumpingWait();
+                    wrapper.WaitForTags();
 
-                var snapshot = buffer.CurrentSnapshot;
-                var spans = tagger.GetTags(new NormalizedSnapshotSpanCollection(new SnapshotSpan(snapshot, 0, snapshot.Length))).ToImmutableArray();
+                    var snapshot = workspace.Documents.First().GetTextBuffer().CurrentSnapshot;
```
Comment (AdamSpeight2008) [inline src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs]: Is this safe to do? What if `Documents` is empty or null?

File: src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs]: This is test code.  And all these tests are written as if they operate on a single document.

**Comment by DustinCampbell** [comment]
:+1:

File: src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs (hunk 2 above)
Comment (jasonmalinowski) [inline src/EditorFeatures/Test/Squiggles/AbstractSquiggleProducerTests.cs]: These are tests, so there will also be at least one Document in them or the test is broken. Documents itself can never be null.

**Pull request merged**